Java fixes, tests and examples for ALS, KMeans

shivaram opened the pull request:
This change addresses some of the issues with using MLLib from Java programs. Specifically it fixes the `train` interface to not use Ints or Doubles and also renames methods to make them easier to call. 

As a way to check if the Java interface works, we have two additions: a. A set of Java examples in the `examples` project that read a data file and run the algorithms. b. A set of unit tests which check if the basic methods can be called and work as expected.

shivaram committed:
Java examples, tests for KMeans and ALS
- Changes ALS to accept RDD[Rating] instead of (Int, Int, Double) making it
  easier to call from Java
- Renames class methods from `train` to `run` to enable static methods to be
  called from Java.
- Add unit tests which check if both static / class methods can be called.
- Also add examples which port the main() function in ALS, KMeans to the
  examples project.

Couple of minor changes to existing code:
- Add a toJavaRDD method in RDD to convert scala RDD to java RDD easily
- Workaround a bug where using double[] from Java leads to class cast exception in
  KMeans init

shivaram committed:
Add a test case for random initialization.
Also workaround a bug where double[][] class cast fails

AmplabJenkins commented:
Thank you for submitting this pull request.

All automated tests for this request have passed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/497/

mateiz commented on examples/src/main/java/spark/mllib/JavaALS.java:
```diff
@@ -0,0 +1,87 @@
+      Integer x = Integer.parseInt(tok.nextToken());
+      Integer y = Integer.parseInt(tok.nextToken());
+      Double rating = Double.parseDouble(tok.nextToken());
+      return new Rating(x, y, rating);
```
I'm curious here, Shivaram, couldn't you pass in primitive types? If that's a problem with Rating being a case class, we should make it not be one, so that it can also be initialized in Java.

shivaram replied:
No there aren't any problems with case classes. I was under the wrong assumption that `Integer.parseInt` returned an `Integer` rather than `int` . I have removed the extra constructor now and changed the types here as well.

mateiz commented on mllib/src/test/scala/spark/mllib/recommendation/JavaALSSuite.java:
```diff
@@ -0,0 +1,110 @@
+
+  @Before
+  public void setUp() {
+    sc = new JavaSparkContext("local", "JavaLogisticRegressionSuite");
```
This should say ALS

mateiz commented:
This looks good, but I just made a couple of comments.

shivaram committed:
Remove Java-specific constructor for Rating.
The scala constructor works for native type java types. Modify examples
to match this.

shivaram commented:
Thanks for the review. Using native types in the case class constructor seems to work fine, so I just deleted the java-specific code.

AmplabJenkins commented:
Thank you for submitting this pull request.

All automated tests for this request have passed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/515/

mateiz commented:
Great, thanks! Going to merge this in.

apivovarov commented:
I think these new 2 classes have wrong path or package

public class  JavaALS
path: examples/src/main/java/spark/mllib/JavaALS.java
package spark.mllib.examples;

public class JavaKMeans
path: examples/src/main/java/spark/mllib/JavaKMeans.java
package spark.mllib.examples;

apivovarov commented:
I also noticed that spark-mllib dependency was not added to hadoop2-yarn profile
so, maven build is brocken

apivovarov commented:
I fixed both issues. PR: https://github.com/mesos/spark/pull/804
